pip/pip-310.md

```diff
@@ -0,0 +1,121 @@
+# PIP-310 Support custom publish rate limiters
+
+# Background knowledge
+
+Pulsar uses rate limiters to control produce and consume rates. At core, this utility is provided by the RateLimiter.java class.
+There are many places where a rate limiter is used today. Mainly:
+* Produce path:
+  * ResourceGroupPublishLimiter - resource group's produce qps and MBps checks
+  * PrecisePublishLimiter - the topic quota rate limiter in case when `preciseTopicPublishRateLimiterEnable` broker config is `true`
```

Comment:
What's the caveat of precise?

Reply:
The caveat should be covered in the Motivation section. Is caveat mentioned there good enough?

Reply:
You wrote:

> The precise one checks for breach before handing out permits each time.

What's the problem with that? calling decrease on an AtomicLong is a performance issue?

Reply:
As I said, its mentioned in the motivation section. Line 38. Quoting here:
> Precise rate limiter fixes the above two issues, but introduces another challenge - the rate limiting is too strict.
>  * This leads potential for data loss in case there are sudden spikes in produce and client side produce queue breaches.
>  * The produce latencies increase exponentially in case produce breaches the set throughput even for small windows.